keras_cv/applications/object_detection/retina_net/retina_net.py

```diff
@@ -0,0 +1,322 @@
+        else:
+            self._metrics_bounding_box_format = self.bounding_box_format
+
+        any_wrong_format = any(
+            [
+                m.bounding_box_format != self._metrics_bounding_box_format
+                for m in metrics
+            ]
+        )
+        if any_wrong_format:
+            raise ValueError(
+                "All metrics passed to RetinaNet.compile() must have "
+                "the same `bounding_box_format` attribute.  For example, if one metric "
+                "uses 'xyxy', all other metrics must use 'xyxy'.  Received "
+                f"metrics={metrics}"
+            )
+
+    def call(self, x, training=False):
+        backbone_outputs = self.backbone(x, training=training)
+        features = self.feature_pyramid(backbone_outputs, training=training)
+
+        N = tf.shape(x)[0]
+        cls_outputs = []
+        box_outputs = []
+        for feature in features:
+            box_outputs.append(tf.reshape(self.box_head(feature), [N, -1, 4]))
+            cls_outputs.append(
+                tf.reshape(self.classification_head(feature), [N, -1, self.num_classes])
+            )
+
+        cls_outputs = tf.concat(cls_outputs, axis=1)
+        box_outputs = tf.concat(box_outputs, axis=1)
+        train_preds = tf.concat([box_outputs, cls_outputs], axis=-1)
+
+        # no-op if default decoder is used.
+        pred_for_inference = bounding_box.convert_format(
+            train_preds,
+            source=self.bounding_box_format,
+            target=self.prediction_decoder.bounding_box_format,
+            images=x,
+        )
+        pred_for_inference = self.prediction_decoder(x, pred_for_inference)
+        pred_for_inference = bounding_box.convert_format(
+            pred_for_inference,
+            source=self.prediction_decoder.bounding_box_format,
+            target=self.bounding_box_format,
+            images=x,
+        )
+        return {"train_preds": train_preds, "inference": pred_for_inference}
+
+    def _encode_data(self, x, y):
+        y_for_metrics = y
+
+        y = bounding_box.convert_format(
+            y,
+            source=self.bounding_box_format,
+            target=self.label_encoder.bounding_box_format,
+            images=x,
+        )
+        y_training_target = self.label_encoder(x, y)
+        y_training_target = bounding_box.convert_format(
+            y_training_target,
+            source=self.label_encoder.bounding_box_format,
+            target=self.bounding_box_format,
+            images=x,
+        )
+        return y_for_metrics, y_training_target
+
+    def train_step(self, data):
+        x, y = data
+        # y comes in in self.bounding_box_format
+        y_for_metrics, y_training_target = self._encode_data(x, y)
+
+        with tf.GradientTape() as tape:
+            predictions = self(x, training=True)
+            # predictions technically do not have a format
+            # loss accepts
+            loss = self.compiled_loss(
```

Comment:
TODO: split loss into classification & box_losses.

This will allow us to export more transparent metrics, as well as have a more modular API.